apps/src/code-studio/components/ExportDialog.jsx

```diff
@@ -206,25 +219,95 @@ class ExportDialog extends React.Component {
   };
 
   componentDidUpdate(prevProps) {
-    const {isOpen, projectUpdatedAt} = this.props;
-    const {exportProjectUpdatedAt} = this.state;
+    const {isOpen, md5SavedSources} = this.props;
+    const {md5PublishSavedSources} = this.state;
     if (isOpen && !prevProps.isOpen) {
       recordExport('open');
-      if (projectUpdatedAt !== exportProjectUpdatedAt) {
-        // The project has changed since we last opened the dialog,
-        // reset our export state, so we will need to export again:
-        this.resetExportState();
+      const md5ApkSavedSources = this.getMd5ApkSavedSources();
+      const sourcesChangedSinceExport =
+        md5ApkSavedSources && md5SavedSources !== md5ApkSavedSources;
+      const sourcesChangedThisInstance =
+        md5PublishSavedSources && md5SavedSources !== md5PublishSavedSources;
+      if (sourcesChangedSinceExport) {
+        // Also cancel a build that may have started before this page was reloaded:
+        this.cancelIfPreexistingApkBuild();
+      }
+      if (sourcesChangedThisInstance) {
+        // The project has changed since we last published within this dialog,
+        // cancel any builds in progress and reset our export state, so we will
+        // start all over again:
+        this.cancelIfGeneratingAndResetState();
       }
     }
   }
 
+  clearWaitTimer() {
+    if (this.waitTimerId) {
+      clearTimeout(this.waitTimerId);
+      this.waitTimerId = null;
+    }
+  }
+
+  cancelIfGeneratingAndResetState() {
+    const {screen, generatingApk} = this.state;
+    if (screen === 'generating' && generatingApk) {
+      this.clearWaitTimer();
+
+      const {exportApp, md5SavedSources} = this.props;
+      const {expoSnackId, apkBuildId} = this.state;
+
+      exportApp({
+        mode: 'expoCancelApkBuild',
+        md5SavedSources,
+        expoSnackId,
+        apkBuildId
+      });
+    }
+    // If we are publishing to expo (state.exporting is true), we don't need
+    // to cancel, just resetting our state is good enough.
+    this.resetExportState();
+  }
+
+  getMd5ApkSavedSources() {
+    const {exportGeneratedProperties = {}} = this.props;
+    const {android = {}} = exportGeneratedProperties;
+    const {md5ApkSavedSources} = android;
```

Comment:
Why not just set md5ApkSavedSources to this.props.exportGeneratedProperties.android? The shorthand is making it difficult to read.

Reply:
Similar comment for 285

Reply:
This allows us to safely dereference without worrying about `undefined` values in the chain.

The other option (if we had the elvis operator enabled in babel) would be

`this.props.exportedGeneratedProperties?.android?.md5ApkSavedSources`

But just doing `this.props.exportGeneratedProperties.android` will cause exceptions when some of those things are undefined (both `exportGeneratedProperties` and `android` are optional so that we don't insist on saving those empty objects into every project)

Reply:
Should we add safe-navigation ("elvis operator"? TIL 😄) to our Babel config? That would make sense as a separate PR to staging, then merged into this PR.

Agree with @jmkulwik that progressively destructuring the object feels like a bit of a hack 😕.

Reply:
I would love adding the elvis operator. I've used that syntax in other projects in other languages and it's very nice. If we don't go that route, change this to the `if (this.props.exportGeneratedProperties && this.props.exportGeneratedProperties.android)` style. It's the style we tend to use in other places.

Reply:
I'll work to add https://babeljs.io/docs/en/babel-plugin-proposal-optional-chaining in a separate PR.

In the meantime, I'll leave this as is until that PR is ready.

Reply:
@jmkulwik @joshlory It looks like we need to move to babel 7 in order to take advantage of this plugin, which is a bigger change.

Also, it may cause issues for linters and editors since the proposal is only at stage 2 (https://github.com/tc39/proposal-optional-chaining). I propose that we revisit this when it is at stage 3.